---
title: "Identifying MYC non-responders in Loven *et al*. RNA-seq"
output: html_notebook
---
This notebook accompanies the analysis described in our manuscript. It details the identification of MYC non-responders (potential housekeeper genes) using RNA-seq data from Loven et al. (2012). These genes will be leveraged to normalize our own RNA-seq data, accounting for MYC's transcriptional amplifier effect.

### Introduction
Reads Per Kilobase per Million mapped reads (RPKM), renormalized based on synthetic spike-ins, were obtained from the GEO database (GSE40784). These samples represent RNA-seq experiments under differing MYC conditions:

-   Low MYC sample: GSM1001393 - P493-6 cells with low c-MYC expression at T=0hr.

-   High MYC sample = GSM1001394 - P493-6 cells with high c-MYC expression at T=24hr.

### Objectives
1. Pre-process and integrate the RNA-seq data from low and high MYC conditions.

2. Identify genes whose expression levels remain stable across MYC conditions (MYC non-responders).

3. Utilize these genes as normalization references for further RNA-seq analyses.

### Load required libraries

```{r}
# load libraries
suppressMessages({
  library(tidyr)
  library(ggplot2)
  library(dplyr)
  library(tibble)
  library(gprofiler2)
  })
```

### Import and Process RNA-seq Data
The spike-in normalized RPKM values for the two samples were downloaded from GEO. Our first step is to process each sample to:

Summarize RPKM values for duplicate gene identifiers: If a gene appears multiple times, we sum its RPKM values.
Combine the two samples into a single data frame: This allows for direct comparison between low and high MYC conditions.

The following function, process_data, is used to automate the first step for each sample:

```{r}
process_data <- function(file, col_name) {
  read.delim(file) %>%
    group_by(Accession) %>%
    summarize(sum_value = sum(.data[[col_name]], na.rm = TRUE)) %>% # Summarize duplicate gene identifiers
    dplyr::rename(SpikeIn = sum_value) # # Rename the summarized column
}
```

### Process Each Sample
We use the process_data function to separately process the high and low MYC samples:

```{r}
# Process both files
high <- process_data("data/GSM1001394_120817_C1250ACXX_7_highMyc.RPKM.txt", "SpikeIn_renormalized_HighMyc")
low <- process_data("data/GSM1001393_120817_C1250ACXX_8_lowMyc.RPKM.txt", "SpikeIn_renormalized_LowMyc")
```

### Merge and Clean Data
Next, we combine the processed samples into a single data frame. Each gene's expression under low and high MYC conditions is aligned, and duplicate rows are removed:

```{r}
# Merge the high and low MYC samples
agg <- inner_join(low, high, by = "Accession") %>% # Merge on 'Accession'
  column_to_rownames(var = "Accession") %>% # Set gene names as rownames
  dplyr::rename(low = SpikeIn.x, high = SpikeIn.y) # Rename columns for clarity

# Clean up environment
rm(high, low) # Remove intermediate objects

```

### Summary Statistics
To understand the distribution of spike-in normalized RPKM values for the two MYC conditions, summary statistics are computed for each sample. These metrics include common measures of central tendency and spread such as the minimum, first quartile, median, mean, third quartile, and maximum.

As expected, the MYC high sample exhibits higher overall RPKM values compared to the MYC low sample. This aligns with findings from the Loven et al. paper, which describe MYC’s role as a transcriptional amplifier, increasing expression levels across a broad range of genes.

```{r}
# Define function 
summarise_sample <- function(data_vector, label) {
  summary_stats <- summary(data_vector)
  summary_df <- data.frame(
    Measure = c(names(summary_stats), "Mean"),
    Value = c(as.vector(summary_stats), mean(data_vector)),
    Sample = label
  )
  return(summary_df)
}

# Apply function to low and high sample
summary_low_df <- summarise_sample(agg$low, "Low")
summary_high_df <- summarise_sample(agg$high, "High")

# Combine
combined_summary_df <- rbind(summary_low_df, summary_high_df)

# Print to console
print(combined_summary_df)
```

### Visualizing Summary Statistics with Facetted ggplot

To better understand and compare the summary statistics of spike-in normalized RPKM values for the MYC low and MYC high samples, the metrics are visualized using a facetted bar chart.

```{r}
# reorder so low appears before high in plots
combined_summary_df$Sample <- factor(combined_summary_df$Sample, levels = c("Low", "High"))

# plot
ggplot(combined_summary_df, aes(x = Sample, y = Value, fill = Sample)) +
  geom_bar(stat = "identity", position = "dodge") +
  facet_wrap(~ Measure, scales = "free_y") +
  labs(fill = "MYC status",
       x = NULL,
       y = "Spike-in normalised RPKM") +
    scale_fill_manual(values = c("RoyalBlue3", "red2")) +
  theme_minimal()
```

### Violin and Histogram Plots of Spike-in Normalized RPKM

To further explore the distribution of spike-in normalised RPKM values, we visualise the data using violin plots and histograms. These plots highlight the differences in global expression levels between the MYC low and MYC high samples, as well as the spread of gene expression values within each condition.

### Violin plot

```{r}
#visualize
#convert data to long format
agg_long <- pivot_longer(agg, cols = c("low", "high"), names_to = "high_or_low", values_to = "RPKM")

# reorder so 'low' appears before 'high' in plots
agg_long$high_or_low <- factor(agg_long$high_or_low, levels = c("low", "high"))

#violin plot - data spread too large
ggplot(agg_long, aes(x = high_or_low, y = RPKM, fill = high_or_low)) +
  geom_violin() +
  labs(fill = "MYC status",
       x = NULL,
       y = "Spike-in normalised RPKM") +
  scale_fill_manual(values = c("RoyalBlue3", "red2")) +
  theme_minimal()
```

### Capped axes

To better highlight differences in mean expression between the MYC conditions, we replot the violin plot with a capped y-axis. Additionally, the mean RPKM for each condition is overlaid as a point.

```{r}
#calculate group means
mean_values <- agg_long %>%
  group_by(high_or_low) %>%
  summarise(mean_RPKM = mean(RPKM))

#violin plot - cap y axis at 100 to help visualization
ggplot(agg_long, aes(x = high_or_low, y = RPKM, fill = high_or_low)) +
  geom_violin() +
  labs(fill = "MYC status",
       x = NULL,
       y = "Spike-in normalised RPKM") +
  scale_fill_manual(values = c("RoyalBlue3", "red2")) +
  stat_summary(fun.y = mean, geom = "point", shape = 3, size = 3, aes(group = high_or_low), color = "black") + # Add mean points
  theme_minimal() +
  coord_cartesian(ylim = c(0, 100))
```

### Histogram

This further highlights differences in the distribution of gene expression levels.The MYC low sample shows a higher proportion of genes with low expression (close to zero RPKM). The MYC high sample exhibits more genes in the higher expression bins, reflecting MYC’s role in upregulating global transcription.

```{r}
#histogram - cap x axis at 20 to help visualization
ggplot(agg_long, aes(x = RPKM, fill = high_or_low)) +
  geom_histogram(binwidth = 1, position = "dodge", color = "black") +
  labs(fill = "MYC status",
       x = "Spike-in normalised RPKM", 
       y = "Frequency") +
  scale_fill_manual(values = c("RoyalBlue3", "red2")) +
  theme_minimal() +
  facet_wrap(~ high_or_low, ncol = 1, labeller = labeller(high_or_low = c(low = "Low", high = "High"))) +
  coord_cartesian(xlim = c(0, 20))
```

### Identifying the Most Stable 2% of Genes

To identify potential 'MYC-resistant' housekeeping genes, we calculate the coefficient of variation (CV) for each gene. This metric highlights genes with consistent expression levels across MYC conditions, as they are less affected by MYC's transcriptional amplifier effect.

1. Filter Lowly Expressed Genes: Genes with a combined expression (sum of low and high RPKM) below 1 are excluded to remove noise.
2. The coefficient of variation (CV) is calculated for each remaining gene
3. Annotate accession numbers with Gene symbols
4. Duplicate gene names are resolved by keeping the most stable entry for each gene.
5. ERCC spike-in standards are excluded
6. Genes are ranked by their CV, and the most stable 2% (lowest CV values) are retained.

```{r}

gene_anno <- read.csv(file="data/gene_annotations.csv")

agg_gene <- agg %>%
  rownames_to_column(var = "GeneID") %>%
  rowwise() %>%
  mutate(sum = sum(c_across(c(low, high)))) %>%  # Calculate the sum of low and high expression
  mutate(coeffvar = sd(c_across(c(low, high))) / mean(c_across(c(low, high))) * 100) %>% # Calculate coefficient of variation
  ungroup() %>%
  filter(sum > 1) %>%  # Remove genes whose joint expression is below 1 RPKM
  inner_join(gene_anno, by = c("GeneID" = "Accession")) %>%  # Merge with gene annotations
  arrange(coeffvar) %>%  # Order by coefficient of variation
  distinct(Gene, .keep_all = TRUE) %>%  # Remove duplicate gene names but keep most stable
  filter(!grepl("ERCC", Gene)) %>%  # Remove ERCC spike-in standards
  filter(coeffvar < quantile(coeffvar, 0.02))  # Keep the most stable 2% of genes

# Print the message with the number of rows
cat("Number of genes to be used as potential housekeepers: ", nrow(agg_gene))

```

### Visualizing Stability with Violin Plots

A violin plot of the remaining genes across the two MYC conditions illustrates their stability. The expression levels of these genes remain consistent, with little variability between the MYC low and MYC high samples.

```{r}
#visualise expression of remaining genes
agg_gene_long <- pivot_longer(agg_gene, cols = c("low", "high"), names_to = "high_or_low", values_to = "RPKM")

# reorder so 'low' appears before 'high' in plots
agg_gene_long$high_or_low <- factor(agg_gene_long$high_or_low, levels = c("low", "high"))

#plot
ggplot(agg_gene_long, aes(x = high_or_low, y = RPKM, fill = high_or_low)) +
  geom_violin() +
  labs(fill = "MYC status",
       x = NULL,
       y = "Spike-in normalised RPKM") +
  scale_fill_manual(values = c("RoyalBlue3", "red2")) +
  theme_minimal() +
  coord_cartesian(ylim = c(0, 100))
```

### Converting to Ensembl IDs and Exporting Results

To facilitate downstream RNA-seq normalization, the candidate genes are mapped to Ensembl Gene IDs (ENSG). These IDs are exported to a CSV file for integration into other pipelines.

```{r}
#convert to ENSG and write to csv.
ensg <- gconvert(agg_gene$Gene, organism = "hsapiens", target = "ENSG") #get ensembl IDs
#write.csv(ensg, file="data/candidate_housekeeper_genes_from_Loven_RNAseq.csv")

#empty environment
#rm(list = ls())
```
